File: lib/motor/score-domain.ts
/**
 * Score domains per asset class.
 *
 * The motor emits two incompatible score scales:
 * - `unit`: cross-sectional percentile rank in [0, 1] (two-layer regime + security models).
 *   0.5 means "median instrument among peers", never "neutral direction".
 * - `signed`: legacy z-score composite in [-1, +1], where the sign is directional.
 *
 * Reading a `unit` score on a signed axis is what produced "Buy gauge + Hold
 * recommendation" for cash instruments, so every consumer must resolve the domain
 * before mapping a score to any Buy/Sell language.
 */

export type ScoreDomain = "unit" | "signed";

export type ClassScoreProfile = {
  domain: ScoreDomain;
  /** Security-layer thresholds (cross-sectional rank) used by the motor. */
  security: { strong: number; weak: number };
  /** Regime-layer thresholds (allocation) used by the motor. */
  regime: { overweight: number; hold: number; reduce: number };
  /**
   * True when instrument prices are NAV-flat by design (cash-like), so momentum
   * oscillators and price-extension targets are not meaningful.
   */
  stabilityFocused: boolean;
  /**
   * How many security-layer indicators the motor is expected to persist. The
   * two-layer models are deliberately parsimonious (cash persists exactly three),
   * so a flat threshold of four would penalize them for working as designed.
   */
  minSecurityIndicators: number;
};

const TWO_LAYER_DEFAULT: ClassScoreProfile = {
  domain: "unit",
  security: { strong: 0.65, weak: 0.25 },
  regime: { overweight: 0.65, hold: 0.45, reduce: 0.25 },
  stabilityFocused: false,
  minSecurityIndicators: 3,
};

const SIGNED_DEFAULT: ClassScoreProfile = {
  domain: "signed",
  security: { strong: 0.3, weak: -0.3 },
  regime: { overweight: 0.3, hold: -0.3, reduce: -0.6 },
  stabilityFocused: false,
  minSecurityIndicators: 4,
};

/** Classes scored by the two-layer regime + security models (motor class_model_registry). */
const TWO_LAYER_CLASSES = [
  "cash_equivalents",
  "fi_treasury",
  "fi_ig",
  "fi_hy",
  "fi_tips",
  "fi_preferred",
  "us_equity",
  "intl_equity",
  "em_equity",
  "real_estate",
  "commodities_precious",
  "commodities_energy",
  "energy_mlp",
  "healthcare_biotech",
  "alt_bdc",
  "alt_infrastructure",
  "currencies",
] as const;

const CLASS_PROFILES: Record<string, ClassScoreProfile> = {
  ...Object.fromEntries(TWO_LAYER_CLASSES.map((id) => [id, TWO_LAYER_DEFAULT])),
  cash_equivalents: { ...TWO_LAYER_DEFAULT, stabilityFocused: true },
};

export function classScoreProfile(classId: string | null | undefined): ClassScoreProfile {
  if (!classId) return SIGNED_DEFAULT;
  return CLASS_PROFILES[classId] ?? SIGNED_DEFAULT;
}

export function scoreDomainForClass(classId: string | null | undefined): ScoreDomain {
  return classScoreProfile(classId).domain;
}

export function isStabilityFocusedClass(classId: string | null | undefined): boolean {
  return classScoreProfile(classId).stabilityFocused;
}

/**
 * Median of a `unit` domain — the point where an instrument is neither better nor
 * worse than its peers. Meaningless for `signed` scores, where 0 is neutral.
 */
export function neutralScore(classId: string | null | undefined): number {
  return scoreDomainForClass(classId) === "unit" ? 0.5 : 0;
}

File: motor/src/decisao/score_domain.py
"""Score domains and thresholds per aba.

Two-layer classes (regime + security models) emit cross-sectional percentile ranks
in [0, 1], where 0.5 means "median instrument among peers". Legacy abas emit signed
z-score composites in [-1, +1], where the sign is directional.

Comparing a unit score against an absolute threshold such as `score > 0.1` is
meaningless: almost every valid rank passes, which is why entry validation used to
approve virtually every cash instrument.
"""

from __future__ import annotations

from typing import Any

from motor.src.config_loader import is_class_model_aba, resolve_aba_config_id

UNIT = "unit"
SIGNED = "signed"

# Percentile bands used by the security models (see *_security_score.py).
UNIT_SECURITY_THRESHOLDS = {"strong": 0.65, "median": 0.5, "weak": 0.25}
SIGNED_SECURITY_THRESHOLDS = {"strong": 0.3, "median": 0.0, "weak": -0.3}

# Allocation bands used by the regime models (see config/models/*_regime.json).
UNIT_REGIME_THRESHOLDS = {"overweight": 0.65, "hold": 0.45, "reduce": 0.25}
SIGNED_REGIME_THRESHOLDS = {"overweight": 0.3, "hold": -0.3, "reduce": -0.6}

# Classes whose instruments are NAV-flat by construction: momentum and price
# extension carry no timing information.
STABILITY_FOCUSED_ABAS = {"cash_equivalents"}


def score_domain_for_aba(aba_id: str) -> str:
    return UNIT if is_class_model_aba(aba_id) else SIGNED


def security_thresholds(aba_id: str) -> dict[str, float]:
    return (
        UNIT_SECURITY_THRESHOLDS
        if score_domain_for_aba(aba_id) == UNIT
        else SIGNED_SECURITY_THRESHOLDS
    )


def regime_thresholds(aba_id: str) -> dict[str, float]:
    return (
        UNIT_REGIME_THRESHOLDS
        if score_domain_for_aba(aba_id) == UNIT
        else SIGNED_REGIME_THRESHOLDS
    )


def is_stability_focused(aba_id: str) -> bool:
    return resolve_aba_config_id(aba_id) in STABILITY_FOCUSED_ABAS


def instrument_quality(aba_id: str, score: float) -> str:
    """Preferred / Competitive / Weak — relative rank, never a Buy/Sell signal."""
    thr = security_thresholds(aba_id)
    if score >= thr["strong"]:
        return "Preferred"
    if score >= thr["weak"]:
        return "Competitive"
    return "Weak"


CANONICAL_ACTIONS = ("Overweight", "Hold", "Reduce", "Strong Reduce")

# Each regime model publishes labels from its own vocabulary (allocation models use
# Overweight/Hold/Reduce, pace models use Acelerar/Manter/Pausar/Reverter). The UI
# reasons about a single 4-value scale, so map them here instead of leaking model
# specific wording into the decision layer.
_ACTION_ALIASES = {
    "overweight": "Overweight",
    "accumulate": "Overweight",
    "acumular": "Overweight",
    "accelerate": "Overweight",
    "acelerar": "Overweight",
    "hold": "Hold",
    "manter": "Hold",
    "base": "Hold",
    "neutral": "Hold",
    "neutro": "Hold",
    "reduce": "Reduce",
    "reduzir": "Reduce",
    "decelerate": "Reduce",
    "desacelerar": "Reduce",
    "pausar": "Reduce",
    "pause": "Reduce",
    "strong reduce": "Strong Reduce",
    "reduzir forte": "Strong Reduce",
    "reverter": "Strong Reduce",
    "reverse": "Strong Reduce",
    "stop": "Strong Reduce",
    "exit": "Strong Reduce",
}


def normalize_allocation_action(action: str | None) -> str | None:
    """Map a model-specific action label onto the canonical 4-value scale."""
    if not action:
        return None
    return _ACTION_ALIASES.get(action.strip().lower())


def allocation_action(aba_id: str, score: float, regime_action: str | None = None) -> str:
    """Overweight / Hold / Reduce / Strong Reduce for the sleeve."""
    normalized = normalize_allocation_action(regime_action)
    if normalized:
        return normalized
    thr = regime_thresholds(aba_id)
    if score >= thr["overweight"]:
        return "Overweight"
    if score >= thr["hold"]:
        return "Hold"
    if score >= thr["reduce"]:
        return "Reduce"
    return "Strong Reduce"


def estagio_to_action(estagio: str | None) -> str:
    mapping = {
        "Ascendente": "Overweight",
        "Maduro": "Hold",
        "Descendente": "Reduce",
        "ForteDescendente": "Strong Reduce",
    }
    return mapping.get(estagio or "", "Hold")


def export_decision_fields(payload: dict[str, Any]) -> dict[str, Any]:
    """Keys consumed by the Next.js decision summary."""
    return {
        "scoreDomain": payload.get("score_domain"),
        "allocationAction": payload.get("allocation_action"),
        "instrumentQuality": payload.get("instrument_quality"),
        "entryTiming": payload.get("entry_timing"),
        "entryReasons": payload.get("entry_reasons", []),
        "peerMedian": payload.get("peer_median"),
    }
